### PR TITLE
Support array validators for multi-select field value

### PR DESCRIPTION
**Pull request by vubogovich**
## What

Support error messages from the validator for the whole array value of multi-select, for example `required` and `min` in the following form validation schema:
```
Yup.object().shape({
  products: Yup.array(
    Yup.string().matches(/^((?!all-the-acid).)*$/, 'Ooops, think again!')
  ).required().min(1, 'Please select at least one product'),
})
```

## Why

Otherwise validation errors lead to the exception or it's not possible to use `required()` or `min()` validation on the array for a multi-select.

## Who is affected

Consumers of multi-selects who already use non-array initial values.

---
This change is [Reviewable](https://reviewable.io/reviews/qualifyze/design-system/160)

### CODE REVIEW

**Review by sirJiggles — CHANGES_REQUESTED** [review]
> Reviewed 1 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/160)* status: 1 of 9 files reviewed, 2 unresolved discussions (waiting on @vubogovich)

___
*[src/components/MultiSelectAsyncCreatableField/index.jsx, line 45 at r1](https://reviewable.io/reviews/qualifyze/design-system/160#-MhCrvxY5UVYv83Ky6UC:-MhCrvxY5UVYv83Ky6UD:b-71825m) ([raw file](https://github.com/qualifyze/design-system/blob/3bca05312f5a4946972c17890fca2ea2d5388e35/src/components/MultiSelectAsyncCreatableField/index.jsx#L45)):*
> ```JavaScript
> 
>   if (hasError) {
>     const errorText = Array.isArray(meta.error) ? meta.error[0] : meta.error
> ```

Are you sure this is ok? I mean that before I had to filter on ! == '' as there is an array where each index maps to each option and error is '' for the ones that do not have an error. At least if I remember correctly, this is why I had the filter to only grab the first one that actually has an error for the option.
___
*[src/components/MultiSelectAsyncCreatableField/index.jsx, line 68 at r1](https://reviewable.io/reviews/qualifyze/design-system/160#-MhCsL_b5yvuBphj5PD6:-MhCsL_b5yvuBphj5PD7:b-sx86y8) ([raw file](https://github.com/qualifyze/design-system/blob/3bca05312f5a4946972c17890fca2ea2d5388e35/src/components/MultiSelectAsyncCreatableField/index.jsx#L68)):*
> ```JavaScript
>           helpers.setValue(valueArray ? valueArray.map(o => o.value) : [])
>           // ...but we need to cache the object array locally
>           setCachedOptions(valueArray || [])
> ```

how come we need the defaulting to empty arrays for the values? sorry I do not get it from the PR descr :see_no_evil: 

I found that required was working, but maybe I was missing something here, maybe there is some interaction I missed?

**Review by gregoralbrecht — CHANGES_REQUESTED** [review]
Can you help me understand what you mean by "Support single error messages"? 

> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/160#-:-MhD3H2a5AnmrPzo-zHP:bgfryuq)* status: 1 of 9 files reviewed, 3 unresolved discussions (waiting on @vubogovich)

*[a discussion](https://reviewable.io/reviews/qualifyze/design-system/160#-MhD3U63COGQ_J1x1CQi:-MhD3U63COGQ_J1x1CQj:b2t343m) (no related file):*
This breaks:

1. https://deploy-preview-160--qualifyze-storybook.netlify.app/?path=/story/multiselectfield--default
2. Select "Ascorbic acid"
3. Select "Don't pick me"

---

**Review by gregoralbrecht — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/160)* status: 1 of 9 files reviewed, 4 unresolved discussions (waiting on @vubogovich)

___
*[src/components/MultiSelectAsyncCreatableField/index.stories.jsx, line 83 at r1](https://reviewable.io/reviews/qualifyze/design-system/160#-MhD4Gbb7_fLeHXcLKBE:-MhD4Gbb7_fLeHXcLKBF:brlwtts) ([raw file](https://github.com/qualifyze/design-system/blob/b60a54471e0db7e5156f13a99c76b8be441034a2/src/components/MultiSelectAsyncCreatableField/index.stories.jsx#L83)):*
> ```JavaScript
>     <Formik
>       initialValues={{
>         products: [],
> ```

As far as I know this isn't possible without React yelling at us. `initialValues` need to be set

**Review by vubogovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/160)* status: 1 of 10 files reviewed, 4 unresolved discussions (waiting on @gregoralbrecht and @sirJiggles)

*[a discussion](https://reviewable.io/reviews/qualifyze/design-system/160#-MhD3U63COGQ_J1x1CQi:-MhDFwVK-I6hKPjwD_cz:b-d9gx3u) (no related file):*
<details><summary><i>Previously, gregoralbrecht (Gregor Albrecht) wrote…</i></summary><blockquote>

This breaks:

1. https://deploy-preview-160--qualifyze-storybook.netlify.app/?path=/story/multiselectfield--default
2. Select "Ascorbic acid"
3. Select "Don't pick me"
</blockquote></details>

Thanks for catching, didn't test with second wrong option :(

---
___
*[src/components/MultiSelectAsyncCreatableField/index.jsx, line 45 at r1](https://reviewable.io/reviews/qualifyze/design-system/160#-MhCrvxY5UVYv83Ky6UC:-MhDG8Xu1CkzdCeprzSz:blxn7er) ([raw file](https://github.com/qualifyze/design-system/blob/3bca05312f5a4946972c17890fca2ea2d5388e35/src/components/MultiSelectAsyncCreatableField/index.jsx#L45)):*
<details><summary><i>Previously, sirJiggles (Gareth Fuller) wrote…</i></summary><blockquote>

Are you sure this is ok? I mean that before I had to filter on ! == '' as there is an array where each index maps to each option and error is '' for the ones that do not have an error. At least if I remember correctly, this is why I had the filter to only grab the first one that actually has an error for the option.
</blockquote></details>

Thanks for catching, didn't test with second wrong option :(
Though I'm getting undefined instead of empty string - updated to cover both cases.
___
*[src/components/MultiSelectAsyncCreatableField/index.jsx, line 68 at r1](https://reviewable.io/reviews/qualifyze/design-system/160#-MhCsL_b5yvuBphj5PD6:-MhDGccgDHk7t3VI9C2H:bgk7567) ([raw file](https://github.com/qualifyze/design-system/blob/3bca05312f5a4946972c17890fca2ea2d5388e35/src/components/MultiSelectAsyncCreatableField/index.jsx#L68)):*
<details><summary><i>Previously, sirJiggles (Gareth Fuller) wrote…</i></summary><blockquote>

how come we need the defaulting to empty arrays for the values? sorry I do not get it from the PR descr :see_no_evil: 

I found that required was working, but maybe I was missing something here, maybe there is some interaction I missed?
</blockquote></details>

`required` for the array field does not seem to consider empty array as missing, only complains on undefined while the Yup docs say it should.
This can be solved by adding a validator for the length of the array but I thought it would be nice to tolerate undefined initial values so that `required` check could also be useful.
The default in this line specifically is just to ensure `cachedOptions` are always defined, otherwise we need to be careful reading it (i.e. `cachedOptions?.find...`).
___
*[src/components/MultiSelectAsyncCreatableField/index.stories.jsx, line 83 at r1](https://reviewable.io/reviews/qualifyze/design-system/160#-MhD4Gbb7_fLeHXcLKBE:-MhDHetuDW1eKEwmkU8N:b-5dy8ba) ([raw file](https://github.com/qualifyze/design-system/blob/b60a54471e0db7e5156f13a99c76b8be441034a2/src/components/MultiSelectAsyncCreatableField/index.stories.jsx#L83)):*
<details><summary><i>Previously, gregoralbrecht (Gregor Albrecht) wrote…</i></summary><blockquote>

As far as I know this isn't possible without React yelling at us. `initialValues` need to be set
</blockquote></details>

`initialValues` is set to an empty object, i.e. with undefined for the multi-select field. I don't need any complains from React.
Though please still challenge the multi-selects tolerance for undefined initial value if you think it's not aligned with the library.

**Review by sirJiggles — APPROVED** [review]
> Reviewed 5 of 9 files at r1, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/160)* status: all files reviewed, 2 unresolved discussions (waiting on @gregoralbrecht)

___
*[src/components/MultiSelectAsyncCreatableField/index.jsx, line 68 at r1](https://reviewable.io/reviews/qualifyze/design-system/160#-MhCsL_b5yvuBphj5PD6:-MhE5jZL2D8VZwbwK4vM:b-nn151r) ([raw file](https://github.com/qualifyze/design-system/blob/3bca05312f5a4946972c17890fca2ea2d5388e35/src/components/MultiSelectAsyncCreatableField/index.jsx#L68)):*
<details><summary><i>Previously, vubogovich (Vladimir Ubogovich) wrote…</i></summary><blockquote>

`required` for the array field does not seem to consider empty array as missing, only complains on undefined while the Yup docs say it should.
This can be solved by adding a validator for the length of the array but I thought it would be nice to tolerate undefined initial values so that `required` check could also be useful.
The default in this line specifically is just to ensure `cachedOptions` are always defined, otherwise we need to be careful reading it (i.e. `cachedOptions?.find...`).
</blockquote></details>


yeah this sounds good, nice!

**Review by gregoralbrecht — CHANGES_REQUESTED** [review]
Okay now I understood. This was not possible before:

```
const multiSelectSchema = Yup.object().shape({
  products: Yup.array(
    Yup.string().matches(/^((?!all-the-acid).)*$/, 'Ooops, think again!')
  )
    .required()
    .min(3, 'at least 3 please'),
})
```

because the `.min(3, 'at least 3 please')` made the select field throw an error. So this PR adds support for field-level validations in Multi(Async|Creatable)Selects :nerd_face: 

Great catch! :clap: 

> Reviewed 6 of 9 files at r1, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/160#-:-MhE9jaD0tj1p_qySOaV:b-rqdqhj)* status: all files reviewed, 1 unresolved discussion (waiting on @vubogovich)

___
*[src/components/MultiSelectAsyncCreatableField/index.stories.jsx, line 83 at r1](https://reviewable.io/reviews/qualifyze/design-system/160#-MhD4Gbb7_fLeHXcLKBE:-MhEAqsaA899tgEKZTBH:b-sjdmdq) ([raw file](https://github.com/qualifyze/design-system/blob/b60a54471e0db7e5156f13a99c76b8be441034a2/src/components/MultiSelectAsyncCreatableField/index.stories.jsx#L83)):*
<details><summary><i>Previously, vubogovich (Vladimir Ubogovich) wrote…</i></summary><blockquote>

`initialValues` is set to an empty object, i.e. with undefined for the multi-select field. I don't need any complains from React.
Though please still challenge the multi-selects tolerance for undefined initial value if you think it's not aligned with the library.
</blockquote></details>

Got it! Seems like React doesn't show that warning anymore, but I do believe we don't need to support undefined initial values. The overhead of making sure we set it correctly seems too big to be worth it. 

Bottom line: I think the assumption that valid initial values are set should still stand.

**Review by vubogovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/160)* status: 3 of 11 files reviewed, 1 unresolved discussion (waiting on @gregoralbrecht, @sirJiggles, and @vubogovich)

___
*[src/components/MultiSelectAsyncCreatableField/index.stories.jsx, line 83 at r1](https://reviewable.io/reviews/qualifyze/design-system/160#-MhD4Gbb7_fLeHXcLKBE:-MhE_Y4a5OScVBVuQ6Lp:b-6ipx6j) ([raw file](https://github.com/qualifyze/design-system/blob/b60a54471e0db7e5156f13a99c76b8be441034a2/src/components/MultiSelectAsyncCreatableField/index.stories.jsx#L83)):*
<details><summary><i>Previously, gregoralbrecht (Gregor Albrecht) wrote…</i></summary><blockquote>

Got it! Seems like React doesn't show that warning anymore, but I do believe we don't need to support undefined initial values. The overhead of making sure we set it correctly seems too big to be worth it. 

Bottom line: I think the assumption that valid initial values are set should still stand.
</blockquote></details>

My point: if we talk about edit form and the initial values coming from the database, nulls and undefined's are often used instead of empty values, so I don't see a good reason forcing DS consumers to default each value if it's quite easy to implement empty === null/undefined.

Yours:
- The assumption that initial values should be defined is coming from [Formik](https://formik.org/docs/api/formik#initialvalues-values) 
```
Even if your form is empty by default, you must initialize all fields with initial values otherwise React will throw an error saying that you have changed an input from uncontrolled to controlled.
```
- I do see your point on paper, but not in practice. Let’s say we didn’t pay attention and sent a PATCH request to our backend with request = { products: '' }, should that not be handled by the backend properly?
- I’m happy to be convinced as soon as we run into this problem! It’s most definitely out of scope for this PR

The latter is very true, so keeping only the fix for multi-select errors.

**Review by vubogovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/160)* status: 3 of 11 files reviewed, 1 unresolved discussion (waiting on @gregoralbrecht, @sirJiggles, and @vubogovich)

___
*[src/components/MultiSelectAsyncCreatableField/index.stories.jsx, line 83 at r1](https://reviewable.io/reviews/qualifyze/design-system/160#-MhD4Gbb7_fLeHXcLKBE:-MhEbWms4vCShlKWU6PP:b-olzsdr) ([raw file](https://github.com/qualifyze/design-system/blob/b60a54471e0db7e5156f13a99c76b8be441034a2/src/components/MultiSelectAsyncCreatableField/index.stories.jsx#L83)):*
<details><summary><i>Previously, vubogovich (Vladimir Ubogovich) wrote…</i></summary><blockquote>

My point: if we talk about edit form and the initial values coming from the database, nulls and undefined's are often used instead of empty values, so I don't see a good reason forcing DS consumers to default each value if it's quite easy to implement empty === null/undefined.

Yours:
- The assumption that initial values should be defined is coming from [Formik](https://formik.org/docs/api/formik#initialvalues-values) 
```
Even if your form is empty by default, you must initialize all fields with initial values otherwise React will throw an error saying that you have changed an input from uncontrolled to controlled.
```
- I do see your point on paper, but not in practice. Let’s say we didn’t pay attention and sent a PATCH request to our backend with request = { products: '' }, should that not be handled by the backend properly?
- I’m happy to be convinced as soon as we run into this problem! It’s most definitely out of scope for this PR

The latter is very true, so keeping only the fix for multi-select errors.
</blockquote></details>

And other minor changes to enforce the correct initial value (array) for multi-select fields.

**Review by gregoralbrecht — APPROVED** [review]
> Reviewed 8 of 8 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/qualifyze/design-system/160)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @vubogovich)
